kvbc/src/categorization/versioned_kv_category.cpp

```diff
@@ -0,0 +1,347 @@
+                                           VersionedOutput &out,
+                                           storage::rocksdb::NativeWriteBatch &batch) {
+  if (updates.empty()) {
+    return;
+  }
+
+  auto hasher = Hasher{};
+  hasher.init();
+  const auto deleted = false;
+  for (auto it = updates.begin(); it != updates.end();) {
+    // Save the iterator as extract() invalidates it.
+    auto extracted_it = it;
+    ++it;
+    auto node = updates.extract(extracted_it);
+    auto &key = node.key();
+    auto &value = node.mapped();
+
+    if (calculate_root_hash) {
+      updateRootHash(key, value.data, hasher);
+    }
+
+    auto versioned_key = VersionedRawKey{std::move(key), block_id};
+
+    updateLatestKeyVersion(versioned_key.value, block_id, batch);
+
+    putValue(versioned_key, VersionedDbValue{DbValue{std::move(value.data)}}, batch);
+
+    addKeyToUpdateInfo(std::move(versioned_key.value), deleted, value.stale_on_update, out);
+  }
+
+  if (calculate_root_hash) {
+    out.root_hash = hasher.finish();
+  }
+}
+
+void VersionedKeyValueCategory::updateLatestKeyVersion(const std::string &key,
+                                                       BlockId version,
+                                                       storage::rocksdb::NativeWriteBatch &batch) {
+  batch.put(latest_ver_cf_, key, serialize(LatestKeyVersion{version}));
+}
+
+void VersionedKeyValueCategory::putValue(const VersionedRawKey &key,
+                                         const VersionedDbValue &value,
+                                         storage::rocksdb::NativeWriteBatch &batch) {
+  batch.put(values_cf_, serialize(key), serialize(value));
+}
+
+void VersionedKeyValueCategory::addKeyToUpdateInfo(std::string &&key,
+                                                   bool deleted,
+                                                   bool stale_on_update,
+                                                   VersionedOutput &out) {
+  out.keys.emplace(std::move(key), VersionedKeyFlags{deleted, stale_on_update});
+}
+
+void VersionedKeyValueCategory::deleteGenesisBlock(BlockId block_id,
+                                                   const VersionedOutput &out,
+                                                   storage::rocksdb::NativeWriteBatch &batch) {
+  for (const auto &[key, flags] : out.keys) {
+    const auto latest_version = getLatestVersion(key);
+    ConcordAssert(latest_version.has_value());
+    if (flags.stale_on_update) {
+      // This key is marked stale-on-update and, therefore, we can remove its value at `block_id`.
+      batch.del(values_cf_, serialize(VersionedRawKey{key, block_id}));
+
+      // If there are no new versions of a key that was marked as stale-on-update at `block_id`, remove the versions
+      // index too.
+      if (*latest_version == block_id) {
+        batch.del(latest_ver_cf_, key);
+      }
+    } else if (*latest_version > block_id) {
```

Comment:
The rationale is that we have a tombstone in the value table and not in the version one and we'd like to preserve the information that this key was deleted. However, taking the approach of https://github.com/vmware/concord-bft/pull/1120 and changing the API, this will change and if the key is deleted and the versions match, we can delete at that point.

Reply:
Yes, you are both correct, I didn't think of the latest version too. Let me fix it!